pkg/ha/ha_service.go

```diff
@@ -61,10 +63,15 @@ func (ha *HAService) setEndpoints(ctx context.Context) error {
 	endpoints := corev1.Endpoints{}
 	err := ha.manager.GetClient().Get(ctx, client.ObjectKey{Namespace: ha.namespace, Name: app.Name}, &endpoints)
 	if err != nil {
-		return fmt.Errorf("updating the service endpoint to point to the new leader: retrieving endpoints: %w", err)
+		if !errors.IsNotFound(err) {
+			return fmt.Errorf("updating the service endpoint to point to the new leader: retrieving endpoints: %w", err)
+		}
+
+		endpoints.ObjectMeta.Namespace = ha.namespace
+		endpoints.ObjectMeta.Name = app.Name
```

Comment:
It seems like there's a chance that the endpoint could not exist. On [line 80](https://github.com/gardener/gardener-custom-metrics/blob/be371fc77f13902aefe0f8ef9fa2ef9b4c9a9d14/pkg/ha/ha_service.go#L80) you use `Update` which will fail if that is the case.

Reply:
It looks like the REST API implements create/update for all verbs for the `endpoints` resource. Per [API reference](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.26/#replace-endpoints-v1-core) all verbs have a `201 CREATED` result (which I didn't expect, but I'm not one to object agains  idempotency).

Anyway, it works nicely for my case - the call to Update() becomes a HTTP PUT, and the server does create the object when it is missing.

CC @ialidzhikov, as you too were interested.

Reply:
Hm, this seems to work only for endpoints but not for pods:
```go
func main() {
	ctx := context.Background()

	restConfig, err := config.GetConfig()
	if err != nil {
		panic(err)
	}

	testClient, err := client.New(restConfig, client.Options{})
	if err != nil {
		panic(err)
	}

	endpoints := corev1.Endpoints{
		ObjectMeta: metav1.ObjectMeta{
			Name:      "foo",
			Namespace: "default",
		},
	}

	endpoints.ObjectMeta.Labels = map[string]string{"app": "foo"}

	if err := testClient.Update(ctx, &endpoints); err != nil {
		panic(err)
	}

	fmt.Println("Successfully updated endpoint foo")
}
```

works for endpoints and creates an endpoints `foo`.

Same logic for pods fails with:
```
% go run main.go
panic: pods "foo" not found
```

To me on high level, this seems to be a design choice for the endpoints API and I don't know of any other K8s resource where Update on non-existing resource creates the resource :)

Reply:
I actually found this out for the `endpoint` rest configurations:https://github.com/kubernetes/kubernetes/blob/a309fadbac3339bc8db9ae0a928a33b8e81ef10f/pkg/registry/core/endpoint/strategy.go#L68-L70

Whereas for pods we have: https://github.com/kubernetes/kubernetes/blob/a309fadbac3339bc8db9ae0a928a33b8e81ef10f/pkg/registry/core/pod/strategy.go#L139-L142

Reply:
Ah, nice dig! 👏